// @pnp/sp-commonjs/webparts/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebPartsPersonalizationScope = exports.WebPartDefinition = exports._WebPartDefinition = exports.WebPartDefinitions = exports._WebPartDefinitions = exports.LimitedWebPartManager = exports._LimitedWebPartManager = void 0;
var tslib_1 = require("tslib");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var odata_1 = require("@pnp/odata-commonjs");
var operations_js_1 = require("../operations.js");
var telemetry_js_1 = require("../telemetry.js");
var _LimitedWebPartManager = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_LimitedWebPartManager, _super);
    function _LimitedWebPartManager() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_LimitedWebPartManager.prototype, "scope", {
        get: function () {
            return telemetry_js_1.tag.configure((0, sharepointqueryable_js_1.SharePointQueryable)(this, "Scope"), "f.scope");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_LimitedWebPartManager.prototype, "webparts", {
        get: function () {
            return (0, exports.WebPartDefinitions)(this, "webparts");
        },
        enumerable: false,
        configurable: true
    });
    _LimitedWebPartManager.prototype.export = function (id) {
        return (0, operations_js_1.spPost)(this.clone(LimitedWebPartManagerCloneFactory, "ExportWebPart"), (0, odata_1.body)({ webPartId: id }));
    };
    _LimitedWebPartManager.prototype.import = function (xml) {
        return (0, operations_js_1.spPost)(this.clone(LimitedWebPartManagerCloneFactory, "ImportWebPart"), (0, odata_1.body)({ webPartXml: xml }));
    };
    return _LimitedWebPartManager;
}(sharepointqueryable_js_1._SharePointQueryable));
exports._LimitedWebPartManager = _LimitedWebPartManager;
var LimitedWebPartManager = function (baseUrl, path) { return new _LimitedWebPartManager(baseUrl, path); };
exports.LimitedWebPartManager = LimitedWebPartManager;
var LimitedWebPartManagerCloneFactory = function (baseUrl, path) { return (0, exports.LimitedWebPartManager)(baseUrl, path); };
var _WebPartDefinitions = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_WebPartDefinitions, _super);
    function _WebPartDefinitions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Gets a web part definition from the collection by id
     *
     * @param id The storage ID of the SPWebPartDefinition to retrieve
     */
    _WebPartDefinitions.prototype.getById = function (id) {
        return (0, exports.WebPartDefinition)(this, "getbyid('" + id + "')");
    };
    /**
     * Gets a web part definition from the collection by storage id
     *
     * @param id The WebPart.ID of the SPWebPartDefinition to retrieve
     */
    _WebPartDefinitions.prototype.getByControlId = function (id) {
        return (0, exports.WebPartDefinition)(this, "getByControlId('" + id + "')");
    };
    return _WebPartDefinitions;
}(sharepointqueryable_js_1._SharePointQueryableCollection));
exports._WebPartDefinitions = _WebPartDefinitions;
exports.WebPartDefinitions = (0, sharepointqueryable_js_1.spInvokableFactory)(_WebPartDefinitions);
var _WebPartDefinition = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_WebPartDefinition, _super);
    function _WebPartDefinition() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_WebPartDefinition.prototype, "webpart", {
        /**
        * Gets the webpart information associated with this definition
        */
        get: function () {
            return (0, sharepointqueryable_js_1.SharePointQueryableInstance)(this, "webpart");
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Saves changes to the Web Part made using other properties and methods on the SPWebPartDefinition object
     */
    _WebPartDefinition.prototype.saveChanges = function () {
        return (0, operations_js_1.spPost)(this.clone(exports.WebPartDefinition, "SaveWebPartChanges"));
    };
    /**
     * Moves the Web Part to a different location on a Web Part Page
     *
     * @param zoneId The ID of the Web Part Zone to which to move the Web Part
     * @param zoneIndex A Web Part zone index that specifies the position at which the Web Part is to be moved within the destination Web Part zone
     */
    _WebPartDefinition.prototype.moveTo = function (zoneId, zoneIndex) {
        return (0, operations_js_1.spPost)(this.clone(exports.WebPartDefinition, "MoveWebPartTo(zoneID='" + zoneId + "', zoneIndex=" + zoneIndex + ")"));
    };
    /**
     * Closes the Web Part. If the Web Part is already closed, this method does nothing
     */
    _WebPartDefinition.prototype.close = function () {
        return (0, operations_js_1.spPost)(this.clone(exports.WebPartDefinition, "CloseWebPart"));
    };
    /**
     * Opens the Web Part. If the Web Part is already closed, this method does nothing
     */
    _WebPartDefinition.prototype.open = function () {
        return (0, operations_js_1.spPost)(this.clone(exports.WebPartDefinition, "OpenWebPart"));
    };
    /**
     * Removes a webpart from a page, all settings will be lost
     */
    _WebPartDefinition.prototype.delete = function () {
        return (0, operations_js_1.spPost)(this.clone(exports.WebPartDefinition, "DeleteWebPart"));
    };
    return _WebPartDefinition;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._WebPartDefinition = _WebPartDefinition;
exports.WebPartDefinition = (0, sharepointqueryable_js_1.spInvokableFactory)(_WebPartDefinition);
var WebPartsPersonalizationScope;
(function (WebPartsPersonalizationScope) {
    WebPartsPersonalizationScope[WebPartsPersonalizationScope["User"] = 0] = "User";
    WebPartsPersonalizationScope[WebPartsPersonalizationScope["Shared"] = 1] = "Shared";
})(WebPartsPersonalizationScope = exports.WebPartsPersonalizationScope || (exports.WebPartsPersonalizationScope = {}));
