// @pnp/sp-commonjs/operations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.spPostDeleteETag = exports.spPostDelete = exports.spPatch = exports.spDelete = exports.spPost = exports.spGet = exports.registerCustomRequestClientFactory = void 0;
var tslib_1 = require("tslib");
var odata_1 = require("@pnp/odata-commonjs");
var sphttpclient_js_1 = require("./sphttpclient.js");
var common_1 = require("@pnp/common-commonjs");
var toabsoluteurl_js_1 = require("./utils/toabsoluteurl.js");
function registerCustomRequestClientFactory(requestClientFactory) {
    httpClientFactory = (0, common_1.isFunc)(requestClientFactory) ? function () { return requestClientFactory; } : defaultFactory;
}
exports.registerCustomRequestClientFactory = registerCustomRequestClientFactory;
var defaultFactory = function (runtime) { return function () { return new sphttpclient_js_1.SPHttpClient(runtime); }; };
var httpClientFactory = defaultFactory;
var send = function (method) {
    return function (o, options) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var runtime, operation, data, batchDependency, url;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        runtime = o.getRuntime();
                        operation = (0, odata_1.defaultPipelineBinder)(httpClientFactory(runtime))(method);
                        data = (0, odata_1.cloneQueryableData)(o.data);
                        batchDependency = (0, common_1.objectDefinedNotNull)(data.batch) ? data.batch.addDependency() : function () {
                            return;
                        };
                        return [4 /*yield*/, (0, toabsoluteurl_js_1.toAbsoluteUrl)(o.toUrlAndQuery(), runtime)];
                    case 1:
                        url = _a.sent();
                        (0, common_1.mergeOptions)(data.options, options);
                        return [2 /*return*/, operation(Object.assign({}, data, {
                                batchDependency: batchDependency,
                                url: url,
                            }))];
                }
            });
        });
    };
};
var spGet = function (o, options) {
    // Fix for #304 - when we clone objects we in some cases then execute a get request
    // in these cases the caching settings were getting dropped from the request
    // this tracks if the object from which this was cloned was caching and applies that to an immediate get request
    // does not affect objects cloned from this as we are using different fields to track the settings so it won't
    // be triggered
    if (o.data.cloneParentWasCaching) {
        o.usingCaching(o.data.cloneParentCacheOptions);
    }
    // if we are forcing caching set that in the data here
    if (o._forceCaching) {
        o.data.useCaching = true;
    }
    return send("GET")(o, options);
};
exports.spGet = spGet;
var spPost = function (o, options) { return send("POST")(o, options); };
exports.spPost = spPost;
var spDelete = function (o, options) { return send("DELETE")(o, options); };
exports.spDelete = spDelete;
var spPatch = function (o, options) { return send("PATCH")(o, options); };
exports.spPatch = spPatch;
var spPostDelete = function (o, options) {
    var opts = Object.assign((0, odata_1.headers)({ "X-HTTP-Method": "DELETE" }), options);
    return (0, exports.spPost)(o, opts);
};
exports.spPostDelete = spPostDelete;
var spPostDeleteETag = function (o, options, eTag) {
    if (eTag === void 0) { eTag = "*"; }
    var opts = Object.assign((0, odata_1.headers)({ "X-HTTP-Method": "DELETE", "IF-Match": eTag }), options);
    return (0, exports.spPost)(o, opts);
};
exports.spPostDeleteETag = spPostDeleteETag;
