// @pnp/sp-commonjs/site-users/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SiteUser = exports._SiteUser = exports.SiteUsers = exports._SiteUsers = void 0;
var tslib_1 = require("tslib");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var types_js_1 = require("../site-groups/types.js");
var common_1 = require("@pnp/common-commonjs");
var metadata_js_1 = require("../utils/metadata.js");
var odata_1 = require("@pnp/odata-commonjs");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
var telemetry_js_1 = require("../telemetry.js");
var _SiteUsers = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_SiteUsers, _super);
    function _SiteUsers() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Gets a user from the collection by id
     *
     * @param id The id of the user to retrieve
     */
    _SiteUsers.prototype.getById = function (id) {
        return telemetry_js_1.tag.configure((0, exports.SiteUser)(this, "getById(" + id + ")"), "sus.getById");
    };
    /**
     * Gets a user from the collection by email
     *
     * @param email The email address of the user to retrieve
     */
    _SiteUsers.prototype.getByEmail = function (email) {
        return telemetry_js_1.tag.configure((0, exports.SiteUser)(this, "getByEmail('" + email + "')"), "sus.getByEmail");
    };
    /**
     * Gets a user from the collection by login name
     *
     * @param loginName The login name of the user to retrieve
     */
    _SiteUsers.prototype.getByLoginName = function (loginName) {
        return telemetry_js_1.tag.configure((0, exports.SiteUser)(this).concat("('!@v::" + encodeURIComponent(loginName) + "')"), "sus.getByLoginName");
    };
    /**
     * Removes a user from the collection by id
     *
     * @param id The id of the user to remove
     */
    _SiteUsers.prototype.removeById = function (id) {
        return (0, operations_js_1.spPost)(this.clone(exports.SiteUsers, "removeById(" + id + ")"));
    };
    /**
     * Removes a user from the collection by login name
     *
     * @param loginName The login name of the user to remove
     */
    _SiteUsers.prototype.removeByLoginName = function (loginName) {
        var o = this.clone(exports.SiteUsers, "removeByLoginName(@v)");
        o.query.set("@v", "'" + encodeURIComponent(loginName) + "'");
        return (0, operations_js_1.spPost)(o);
    };
    /**
     * Adds a user to a site collection
     *
     * @param loginName The login name of the user to add  to a site collection
     *
     */
    _SiteUsers.prototype.add = function (loginName) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.spPost)(this, (0, odata_1.body)((0, common_1.assign)((0, metadata_js_1.metadata)("SP.User"), { LoginName: loginName })))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, this.getByLoginName(loginName)];
                }
            });
        });
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("sus.remId")
    ], _SiteUsers.prototype, "removeById", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("sus.remLoginName")
    ], _SiteUsers.prototype, "removeByLoginName", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("sus.add")
    ], _SiteUsers.prototype, "add", null);
    _SiteUsers = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("siteusers")
    ], _SiteUsers);
    return _SiteUsers;
}(sharepointqueryable_js_1._SharePointQueryableCollection));
exports._SiteUsers = _SiteUsers;
exports.SiteUsers = (0, sharepointqueryable_js_1.spInvokableFactory)(_SiteUsers);
/**
 * Describes a single user
 *
 */
var _SiteUser = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_SiteUser, _super);
    function _SiteUser() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.delete = (0, sharepointqueryable_js_1.deleteable)("su");
        /**
        * Updates this user instance with the supplied properties
        *
        * @param properties A plain object of property names and values to update for the user
        */
        _this.update = _this._update("SP.User", function (data) { return ({ data: data, user: _this }); });
        return _this;
    }
    Object.defineProperty(_SiteUser.prototype, "groups", {
        /**
         * Gets the groups for this user
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, types_js_1.SiteGroups)(this, "groups"), "su.groups");
        },
        enumerable: false,
        configurable: true
    });
    return _SiteUser;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._SiteUser = _SiteUser;
exports.SiteUser = (0, sharepointqueryable_js_1.spInvokableFactory)(_SiteUser);
