// @pnp/sp-commonjs/navigation/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NavigationService = exports._NavigationService = exports.Navigation = exports._Navigation = exports.NavigationNode = exports._NavigationNode = exports.NavigationNodes = exports._NavigationNodes = void 0;
var tslib_1 = require("tslib");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var common_1 = require("@pnp/common-commonjs");
var metadata_js_1 = require("../utils/metadata.js");
var odata_1 = require("@pnp/odata-commonjs");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
var telemetry_js_1 = require("../telemetry.js");
/**
 * Represents a collection of navigation nodes
 *
 */
var _NavigationNodes = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_NavigationNodes, _super);
    function _NavigationNodes() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Gets a navigation node by id
     *
     * @param id The id of the node
     */
    _NavigationNodes.prototype.getById = function (id) {
        return telemetry_js_1.tag.configure((0, exports.NavigationNode)(this).concat("(" + id + ")"), "nns.getById");
    };
    /**
     * Adds a new node to the collection
     *
     * @param title Display name of the node
     * @param url The url of the node
     * @param visible If true the node is visible, otherwise it is hidden (default: true)
     */
    _NavigationNodes.prototype.add = function (title, url, visible) {
        if (visible === void 0) { visible = true; }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var postBody, data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        postBody = (0, odata_1.body)((0, common_1.assign)((0, metadata_js_1.metadata)("SP.NavigationNode"), {
                            IsVisible: visible,
                            Title: title,
                            Url: url,
                        }));
                        return [4 /*yield*/, (0, operations_js_1.spPost)(this.clone(exports.NavigationNodes, null), postBody)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                node: this.getById(data.Id),
                            }];
                }
            });
        });
    };
    /**
     * Moves a node to be after another node in the navigation
     *
     * @param nodeId Id of the node to move
     * @param previousNodeId Id of the node after which we move the node specified by nodeId
     */
    _NavigationNodes.prototype.moveAfter = function (nodeId, previousNodeId) {
        var postBody = (0, odata_1.body)({
            nodeId: nodeId,
            previousNodeId: previousNodeId,
        });
        return (0, operations_js_1.spPost)(this.clone(exports.NavigationNodes, "MoveAfter"), postBody);
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("nns.add")
    ], _NavigationNodes.prototype, "add", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("nns.moveAfter")
    ], _NavigationNodes.prototype, "moveAfter", null);
    return _NavigationNodes;
}(sharepointqueryable_js_1._SharePointQueryableCollection));
exports._NavigationNodes = _NavigationNodes;
exports.NavigationNodes = (0, sharepointqueryable_js_1.spInvokableFactory)(_NavigationNodes);
/**
 * Represents an instance of a navigation node
 *
 */
var _NavigationNode = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_NavigationNode, _super);
    function _NavigationNode() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.delete = (0, sharepointqueryable_js_1.deleteable)("nn");
        return _this;
    }
    Object.defineProperty(_NavigationNode.prototype, "children", {
        /**
         * Represents the child nodes of this node
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, exports.NavigationNodes)(this, "children"), "nn.children");
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Updates this node
     *
     * @param properties Properties used to update this node
     */
    _NavigationNode.prototype.update = function (properties) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var postBody, data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        postBody = (0, odata_1.body)((0, common_1.assign)((0, metadata_js_1.metadata)("SP.NavigationNode"), properties), (0, odata_1.headers)({ "X-HTTP-Method": "MERGE" }));
                        return [4 /*yield*/, (0, operations_js_1.spPost)(this, postBody)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                node: this,
                            }];
                }
            });
        });
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("nn.update")
    ], _NavigationNode.prototype, "update", null);
    return _NavigationNode;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._NavigationNode = _NavigationNode;
exports.NavigationNode = (0, sharepointqueryable_js_1.spInvokableFactory)(_NavigationNode);
/**
 * Exposes the navigation components
 *
 */
var _Navigation = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Navigation, _super);
    function _Navigation() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_Navigation.prototype, "quicklaunch", {
        /**
         * Gets the quicklaunch navigation nodes for the current context
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, exports.NavigationNodes)(this, "quicklaunch"), "n.quicklaunch");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Navigation.prototype, "topNavigationBar", {
        /**
         * Gets the top bar navigation nodes for the current context
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, exports.NavigationNodes)(this, "topnavigationbar"), "n.topnavigationbar");
        },
        enumerable: false,
        configurable: true
    });
    _Navigation = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("navigation")
    ], _Navigation);
    return _Navigation;
}(sharepointqueryable_js_1._SharePointQueryable));
exports._Navigation = _Navigation;
exports.Navigation = (0, sharepointqueryable_js_1.spInvokableFactory)(_Navigation);
/**
 * Represents the top level navigation service
 */
var _NavigationService = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_NavigationService, _super);
    function _NavigationService(path) {
        if (path === void 0) { path = null; }
        return _super.call(this, "_api/navigation", path) || this;
    }
    /**
     * The MenuState service operation returns a Menu-State (dump) of a SiteMapProvider on a site.
     *
     * @param menuNodeKey MenuNode.Key of the start node within the SiteMapProvider If no key is provided the SiteMapProvider.RootNode will be the root of the menu state.
     * @param depth Depth of the dump. If no value is provided a dump with the depth of 10 is returned
     * @param mapProviderName The name identifying the SiteMapProvider to be used
     * @param customProperties comma seperated list of custom properties to be returned.
     */
    _NavigationService.prototype.getMenuState = function (menuNodeKey, depth, mapProviderName, customProperties) {
        if (menuNodeKey === void 0) { menuNodeKey = null; }
        if (depth === void 0) { depth = 10; }
        if (mapProviderName === void 0) { mapProviderName = null; }
        if (customProperties === void 0) { customProperties = null; }
        return (0, operations_js_1.spPost)((0, exports.NavigationService)("MenuState"), (0, odata_1.body)({
            customProperties: customProperties,
            depth: depth,
            mapProviderName: mapProviderName,
            menuNodeKey: menuNodeKey,
        }));
    };
    /**
     * Tries to get a SiteMapNode.Key for a given URL within a site collection.
     *
     * @param currentUrl A url representing the SiteMapNode
     * @param mapProviderName The name identifying the SiteMapProvider to be used
     */
    _NavigationService.prototype.getMenuNodeKey = function (currentUrl, mapProviderName) {
        if (mapProviderName === void 0) { mapProviderName = null; }
        return (0, operations_js_1.spPost)((0, exports.NavigationService)("MenuNodeKey"), (0, odata_1.body)({
            currentUrl: currentUrl,
            mapProviderName: mapProviderName,
        }));
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("ns.getMenuState")
    ], _NavigationService.prototype, "getMenuState", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("ns.getMenuNodeKey")
    ], _NavigationService.prototype, "getMenuNodeKey", null);
    return _NavigationService;
}(sharepointqueryable_js_1._SharePointQueryable));
exports._NavigationService = _NavigationService;
var NavigationService = function (path) { return new _NavigationService(path); };
exports.NavigationService = NavigationService;
