// @pnp/sp-commonjs/security/item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var odata_1 = require("@pnp/odata-commonjs");
var types_js_1 = require("../items/types.js");
var types_js_2 = require("./types.js");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var funcs_js_1 = require("./funcs.js");
(0, odata_1.addProp)(types_js_1._Item, "roleAssignments", types_js_2.RoleAssignments);
(0, odata_1.addProp)(types_js_1._Item, "firstUniqueAncestorSecurableObject", sharepointqueryable_js_1.SharePointQueryableInstance);
types_js_1._Item.prototype.getUserEffectivePermissions = funcs_js_1.getUserEffectivePermissions;
types_js_1._Item.prototype.getCurrentUserEffectivePermissions = funcs_js_1.getCurrentUserEffectivePermissions;
types_js_1._Item.prototype.breakRoleInheritance = funcs_js_1.breakRoleInheritance;
types_js_1._Item.prototype.resetRoleInheritance = funcs_js_1.resetRoleInheritance;
types_js_1._Item.prototype.userHasPermissions = funcs_js_1.userHasPermissions;
types_js_1._Item.prototype.currentUserHasPermissions = funcs_js_1.currentUserHasPermissions;
types_js_1._Item.prototype.hasPermissions = funcs_js_1.hasPermissions;
