// @pnp/sp-commonjs/security/web.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var odata_1 = require("@pnp/odata-commonjs");
var types_js_1 = require("../webs/types.js");
var types_js_2 = require("./types.js");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var funcs_js_1 = require("./funcs.js");
(0, odata_1.addProp)(types_js_1._Web, "roleDefinitions", types_js_2.RoleDefinitions);
(0, odata_1.addProp)(types_js_1._Web, "roleAssignments", types_js_2.RoleAssignments);
(0, odata_1.addProp)(types_js_1._Web, "firstUniqueAncestorSecurableObject", sharepointqueryable_js_1.SharePointQueryableInstance);
types_js_1._Web.prototype.getUserEffectivePermissions = funcs_js_1.getUserEffectivePermissions;
types_js_1._Web.prototype.getCurrentUserEffectivePermissions = funcs_js_1.getCurrentUserEffectivePermissions;
types_js_1._Web.prototype.breakRoleInheritance = funcs_js_1.breakRoleInheritance;
types_js_1._Web.prototype.resetRoleInheritance = funcs_js_1.resetRoleInheritance;
types_js_1._Web.prototype.userHasPermissions = funcs_js_1.userHasPermissions;
types_js_1._Web.prototype.currentUserHasPermissions = funcs_js_1.currentUserHasPermissions;
types_js_1._Web.prototype.hasPermissions = funcs_js_1.hasPermissions;
