// @pnp/sp-commonjs/taxonomy/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Terms = exports.Children = exports.Term = exports.Relations = exports.Relation = exports.TermSet = exports.TermSets = exports.TermGroups = exports.TermGroup = exports.TermStore = void 0;
var rest_js_1 = require("../rest.js");
var types_js_1 = require("./types.js");
var types_js_2 = require("./types.js");
Object.defineProperty(exports, "TermStore", { enumerable: true, get: function () { return types_js_2.TermStore; } });
Object.defineProperty(exports, "TermGroup", { enumerable: true, get: function () { return types_js_2.TermGroup; } });
Object.defineProperty(exports, "TermGroups", { enumerable: true, get: function () { return types_js_2.TermGroups; } });
Object.defineProperty(exports, "TermSets", { enumerable: true, get: function () { return types_js_2.TermSets; } });
Object.defineProperty(exports, "TermSet", { enumerable: true, get: function () { return types_js_2.TermSet; } });
Object.defineProperty(exports, "Relation", { enumerable: true, get: function () { return types_js_2.Relation; } });
Object.defineProperty(exports, "Relations", { enumerable: true, get: function () { return types_js_2.Relations; } });
Object.defineProperty(exports, "Term", { enumerable: true, get: function () { return types_js_2.Term; } });
Object.defineProperty(exports, "Children", { enumerable: true, get: function () { return types_js_2.Children; } });
Object.defineProperty(exports, "Terms", { enumerable: true, get: function () { return types_js_2.Terms; } });
Reflect.defineProperty(rest_js_1.SPRest.prototype, "termStore", {
    configurable: true,
    enumerable: true,
    get: function () {
        return this.childConfigHook(function (_a) {
            var options = _a.options, baseUrl = _a.baseUrl, runtime = _a.runtime;
            return (0, types_js_1.TermStore)(baseUrl).configure(options).setRuntime(runtime);
        });
    },
});
