// @pnp/sp-commonjs/comments/clientside-page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var types_js_1 = require("../clientside-pages/types.js");
var types_js_2 = require("./types.js");
var odata_js_1 = require("../odata.js");
var telemetry_js_1 = require("../telemetry.js");
var index_js_1 = require("../items/index.js");
var operations_js_1 = require("../operations.js");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
types_js_1._ClientsidePage.prototype.addComment = function (info) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.comments.add(info)];
            }
        });
    });
};
types_js_1._ClientsidePage.prototype.getCommentById = function (id) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.comments.getById(id).usingParser((0, odata_js_1.spODataEntity)(types_js_2.Comment))()];
            }
        });
    });
};
types_js_1._ClientsidePage.prototype.clearComments = function () {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.comments.clear()];
            }
        });
    });
};
types_js_1._ClientsidePage.prototype.getComments = function () {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, telemetry_js_1.tag.configure(item, "").comments()];
            }
        });
    });
};
types_js_1._ClientsidePage.prototype.like = function () {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem("ID")];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, (0, operations_js_1.spPost)((0, sharepointqueryable_js_1.SharePointQueryable)(item, "like"))];
            }
        });
    });
};
types_js_1._ClientsidePage.prototype.unlike = function () {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem("ID")];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, (0, operations_js_1.spPost)((0, sharepointqueryable_js_1.SharePointQueryable)(item, "unlike"))];
            }
        });
    });
};
types_js_1._ClientsidePage.prototype.getLikedByInformation = function () {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem("ID")];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.getLikedByInformation()];
            }
        });
    });
};
types_js_1._ClientsidePage.prototype.enableComments = function () {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var _this = this;
        return (0, tslib_1.__generator)(this, function (_a) {
            telemetry_js_1.tag.configure(this, "csp.enableComments");
            return [2 /*return*/, this.setCommentsOn(true).then(function (r) {
                    _this.commentsDisabled = false;
                    return r;
                })];
        });
    });
};
types_js_1._ClientsidePage.prototype.disableComments = function () {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var _this = this;
        return (0, tslib_1.__generator)(this, function (_a) {
            telemetry_js_1.tag.configure(this, "csp.disableComments");
            return [2 /*return*/, this.setCommentsOn(false).then(function (r) {
                    _this.commentsDisabled = true;
                    return r;
                })];
        });
    });
};
types_js_1._ClientsidePage.prototype.setCommentsOn = function (on) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, (0, index_js_1.Item)(item, "SetCommentsDisabled(" + !on + ")").update({})];
            }
        });
    });
};
