// @pnp/sp-commonjs/sputilities/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utilities = exports._Utilities = void 0;
var tslib_1 = require("tslib");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var common_1 = require("@pnp/common-commonjs");
var odata_1 = require("@pnp/odata-commonjs");
var odata_js_1 = require("../odata.js");
var metadata_js_1 = require("../utils/metadata.js");
var types_js_1 = require("../files/types.js");
var extractweburl_js_1 = require("../utils/extractweburl.js");
var operations_js_1 = require("../operations.js");
var telemetry_js_1 = require("../telemetry.js");
var objectToSPKeyValueCollection_js_1 = require("../utils/objectToSPKeyValueCollection.js");
var _Utilities = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Utilities, _super);
    function _Utilities(baseUrl, methodName) {
        var _this = this;
        var url = typeof baseUrl === "string" ? baseUrl : baseUrl.toUrl();
        _this = _super.call(this, (0, extractweburl_js_1.extractWebUrl)(url), "_api/SP.Utilities.Utility." + methodName) || this;
        return _this;
    }
    _Utilities.prototype.excute = function (props) {
        return (0, operations_js_1.spPost)(this, (0, odata_1.body)(props));
    };
    _Utilities.prototype.sendEmail = function (props) {
        var params = {
            properties: (0, common_1.assign)((0, metadata_js_1.metadata)("SP.Utilities.EmailProperties"), {
                Body: props.Body,
                From: props.From,
                Subject: props.Subject,
            }),
        };
        if (props.To && props.To.length > 0) {
            params.properties = (0, common_1.assign)(params.properties, {
                To: { results: props.To },
            });
        }
        if (props.CC && props.CC.length > 0) {
            params.properties = (0, common_1.assign)(params.properties, {
                CC: { results: props.CC },
            });
        }
        if (props.BCC && props.BCC.length > 0) {
            params.properties = (0, common_1.assign)(params.properties, {
                BCC: { results: props.BCC },
            });
        }
        if (props.AdditionalHeaders) {
            params.properties = (0, common_1.assign)(params.properties, {
                AdditionalHeaders: (0, objectToSPKeyValueCollection_js_1.objectToSPKeyValueCollection)(props.AdditionalHeaders),
            });
        }
        return telemetry_js_1.tag.configure(this.clone(UtilitiesCloneFactory, "SendEmail", true), "u.sendEmail").excute(params);
    };
    _Utilities.prototype.getCurrentUserEmailAddresses = function () {
        return telemetry_js_1.tag.configure(this.clone(UtilitiesCloneFactory, "GetCurrentUserEmailAddresses", true), "u.getCurrentUserEmailAddresses").excute({});
    };
    _Utilities.prototype.resolvePrincipal = function (input, scopes, sources, inputIsEmailOnly, addToUserInfoList, matchUserInfoList) {
        if (matchUserInfoList === void 0) { matchUserInfoList = false; }
        var params = {
            addToUserInfoList: addToUserInfoList,
            input: input,
            inputIsEmailOnly: inputIsEmailOnly,
            matchUserInfoList: matchUserInfoList,
            scopes: scopes,
            sources: sources,
        };
        var clone = this.clone(UtilitiesCloneFactory, "ResolvePrincipalInCurrentContext", true);
        return telemetry_js_1.tag.configure(clone, "u.ResolvePrincipalInCurrentContext").excute(params);
    };
    _Utilities.prototype.searchPrincipals = function (input, scopes, sources, groupName, maxCount) {
        var params = {
            groupName: groupName,
            input: input,
            maxCount: maxCount,
            scopes: scopes,
            sources: sources,
        };
        var clone = this.clone(UtilitiesCloneFactory, "SearchPrincipalsUsingContextWeb", true);
        return telemetry_js_1.tag.configure(clone, "u.SearchPrincipalsUsingContextWeb").excute(params);
    };
    _Utilities.prototype.createEmailBodyForInvitation = function (pageAddress) {
        var params = {
            pageAddress: pageAddress,
        };
        var clone = this.clone(UtilitiesCloneFactory, "CreateEmailBodyForInvitation", true);
        return telemetry_js_1.tag.configure(clone, "u.CreateEmailBodyForInvitation").excute(params);
    };
    _Utilities.prototype.expandGroupsToPrincipals = function (inputs, maxCount) {
        if (maxCount === void 0) { maxCount = 30; }
        var params = {
            inputs: inputs,
            maxCount: maxCount,
        };
        var clone = this.clone(UtilitiesCloneFactory, "ExpandGroupsToPrincipals", true);
        return telemetry_js_1.tag.configure(clone, "u.ExpandGroupsToPrincipals").excute(params);
    };
    _Utilities.prototype.createWikiPage = function (info) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var clone, newPage;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        clone = this.clone(UtilitiesCloneFactory, "CreateWikiPageInContextWeb", true);
                        return [4 /*yield*/, telemetry_js_1.tag.configure(clone, "u.CreateWikiPageInContextWeb").excute({ parameters: info })];
                    case 1:
                        newPage = _a.sent();
                        return [2 /*return*/, {
                                data: newPage,
                                file: (0, types_js_1.File)((0, odata_js_1.odataUrlFrom)(newPage)),
                            }];
                }
            });
        });
    };
    return _Utilities;
}(sharepointqueryable_js_1._SharePointQueryable));
exports._Utilities = _Utilities;
exports.Utilities = (0, sharepointqueryable_js_1.spInvokableFactory)(_Utilities);
var UtilitiesCloneFactory = function (baseUrl, path) { return (0, exports.Utilities)(baseUrl, path); };
