// @pnp/sp-commonjs/social/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocialStatusCode = exports.SocialFollowResult = exports.SocialActorTypes = exports.SocialActorType = exports.Social = exports.MySocial = void 0;
var types_js_1 = require("./types.js");
var rest_js_1 = require("../rest.js");
var types_js_2 = require("./types.js");
Object.defineProperty(exports, "MySocial", { enumerable: true, get: function () { return types_js_2.MySocial; } });
Object.defineProperty(exports, "Social", { enumerable: true, get: function () { return types_js_2.Social; } });
Object.defineProperty(exports, "SocialActorType", { enumerable: true, get: function () { return types_js_2.SocialActorType; } });
Object.defineProperty(exports, "SocialActorTypes", { enumerable: true, get: function () { return types_js_2.SocialActorTypes; } });
Object.defineProperty(exports, "SocialFollowResult", { enumerable: true, get: function () { return types_js_2.SocialFollowResult; } });
Object.defineProperty(exports, "SocialStatusCode", { enumerable: true, get: function () { return types_js_2.SocialStatusCode; } });
Reflect.defineProperty(rest_js_1.SPRest.prototype, "social", {
    configurable: true,
    enumerable: true,
    get: function () {
        return this.childConfigHook(function (_a) {
            var options = _a.options, baseUrl = _a.baseUrl, runtime = _a.runtime;
            return (0, types_js_1.Social)(baseUrl).configure(options).setRuntime(runtime);
        });
    },
});
