// @pnp/sp-commonjs/folders/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Folder = exports._Folder = exports.Folders = exports._Folders = void 0;
var tslib_1 = require("tslib");
var common_1 = require("@pnp/common-commonjs");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var odata_js_1 = require("../odata.js");
var types_js_1 = require("../items/types.js");
var odata_1 = require("@pnp/odata-commonjs");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
var escapeQueryStrValue_js_1 = require("../utils/escapeQueryStrValue.js");
var extractweburl_js_1 = require("../utils/extractweburl.js");
var telemetry_js_1 = require("../telemetry.js");
var toResourcePath_js_1 = require("../utils/toResourcePath.js");
var _Folders = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Folders, _super);
    function _Folders() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Gets a folder by it's name
     *
     * @param name Folder's name
     */
    _Folders.prototype.getByName = function (name) {
        return telemetry_js_1.tag.configure((0, exports.Folder)(this).concat("('" + (0, escapeQueryStrValue_js_1.escapeQueryStrValue)(name) + "')"), "fs.getByName");
    };
    /**
     * Adds a new folder at the specified URL
     *
     * @param url
     */
    _Folders.prototype.add = function (url) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.spPost)(this.clone(exports.Folders, "add('" + (0, escapeQueryStrValue_js_1.escapeQueryStrValue)(url) + "')"))];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                folder: this.getByName(url),
                            }];
                }
            });
        });
    };
    /**
     * Adds a new folder by path and should be prefered over add
     *
     * @param serverRelativeUrl The server relative url of the new folder to create
     * @param overwrite True to overwrite an existing folder, default false
     */
    _Folders.prototype.addUsingPath = function (serverRelativeUrl, overwrite) {
        if (overwrite === void 0) { overwrite = false; }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.spPost)(this.clone(exports.Folders, "addUsingPath(DecodedUrl='" + (0, escapeQueryStrValue_js_1.escapeQueryStrValue)(serverRelativeUrl) + "',overwrite=" + overwrite + ")"))];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                folder: (0, exports.Folder)((0, extractweburl_js_1.extractWebUrl)(this.toUrl()), "_api/web/getFolderByServerRelativePath(decodedUrl='" + (0, escapeQueryStrValue_js_1.escapeQueryStrValue)(serverRelativeUrl) + "')"),
                            }];
                }
            });
        });
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("fs.add")
    ], _Folders.prototype, "add", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("fs.addUsingPath")
    ], _Folders.prototype, "addUsingPath", null);
    _Folders = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("folders")
    ], _Folders);
    return _Folders;
}(sharepointqueryable_js_1._SharePointQueryableCollection));
exports._Folders = _Folders;
exports.Folders = (0, sharepointqueryable_js_1.spInvokableFactory)(_Folders);
var _Folder = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Folder, _super);
    function _Folder() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.delete = (0, sharepointqueryable_js_1.deleteableWithETag)("f");
        /**
         * Updates folder's properties
         * @param props Folder's properties to update
         */
        _this.update = _this._update("SP.Folder", function (data) { return ({ data: data, folder: _this }); });
        return _this;
    }
    Object.defineProperty(_Folder.prototype, "contentTypeOrder", {
        /**
         * Specifies the sequence in which content types are displayed.
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, sharepointqueryable_js_1.SharePointQueryableCollection)(this, "contentTypeOrder"), "f.contentTypeOrder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Folder.prototype, "folders", {
        /**
         * Gets this folder's sub folders
         *
         */
        get: function () {
            return (0, exports.Folders)(this);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Folder.prototype, "listItemAllFields", {
        /**
         * Gets this folder's list item field values
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, sharepointqueryable_js_1.SharePointQueryableInstance)(this, "listItemAllFields"), "f.listItemAllFields");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Folder.prototype, "parentFolder", {
        /**
         * Gets the parent folder, if available
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, exports.Folder)(this, "parentFolder"), "f.parentFolder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Folder.prototype, "properties", {
        /**
         * Gets this folder's properties
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, sharepointqueryable_js_1.SharePointQueryableInstance)(this, "properties"), "f.properties");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Folder.prototype, "serverRelativeUrl", {
        /**
         * Gets this folder's server relative url
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, sharepointqueryable_js_1.SharePointQueryable)(this, "serverRelativeUrl"), "f.serverRelativeUrl");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(_Folder.prototype, "uniqueContentTypeOrder", {
        /**
         * Gets a value that specifies the content type order.
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, sharepointqueryable_js_1.SharePointQueryableCollection)(this, "uniqueContentTypeOrder"), "f.uniqueContentTypeOrder");
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Moves the folder to the Recycle Bin and returns the identifier of the new Recycle Bin item.
     */
    _Folder.prototype.recycle = function () {
        return (0, operations_js_1.spPost)(this.clone(exports.Folder, "recycle"));
    };
    /**
     * Gets the associated list item for this folder, loading the default properties
     */
    _Folder.prototype.getItem = function () {
        var selects = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            selects[_i] = arguments[_i];
        }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var q;
            var _a;
            return (0, tslib_1.__generator)(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, (_a = this.listItemAllFields).select.apply(_a, (0, tslib_1.__spreadArray)([], (0, tslib_1.__read)(selects), false))()];
                    case 1:
                        q = _b.sent();
                        if ((0, common_1.hOP)(q, "odata.null") && q["odata.null"]) {
                            throw Error("No associated item was found for this folder. It may be the root folder, which does not have an item.");
                        }
                        return [2 /*return*/, (0, common_1.assign)((0, types_js_1.Item)((0, odata_js_1.odataUrlFrom)(q)).configureFrom(this), q)];
                }
            });
        });
    };
    /**
     * Moves a folder to destination path
     *
     * @param destUrl Absolute or relative URL of the destination path
     */
    _Folder.prototype.moveTo = function (destUrl) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var urlInfo, uri;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getParentInfos()];
                    case 1:
                        urlInfo = _a.sent();
                        uri = new URL(urlInfo.ParentWeb.Url);
                        return [4 /*yield*/, (0, operations_js_1.spPost)((0, exports.Folder)(urlInfo.ParentWeb.Url, "/_api/SP.MoveCopyUtil.MoveFolder()").configureFrom(this), (0, odata_1.body)({
                                destUrl: (0, common_1.isUrlAbsolute)(destUrl) ? destUrl : (0, common_1.combine)(uri.origin, destUrl),
                                srcUrl: (0, common_1.combine)(uri.origin, urlInfo.Folder.ServerRelativeUrl),
                            }))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Moves a folder by path to destination path
     * Also works with different site collections.
     *
     * @param destUrl Absolute or relative URL of the destination path
     * @param keepBoth Keep both if folder with the same name in the same location already exists?
     */
    _Folder.prototype.moveByPath = function (destUrl, KeepBoth) {
        if (KeepBoth === void 0) { KeepBoth = false; }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var urlInfo, uri;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getParentInfos()];
                    case 1:
                        urlInfo = _a.sent();
                        uri = new URL(urlInfo.ParentWeb.Url);
                        return [4 /*yield*/, (0, operations_js_1.spPost)((0, exports.Folder)(uri.origin, "/_api/SP.MoveCopyUtil.MoveFolderByPath()").configureFrom(this), (0, odata_1.body)({
                                destPath: (0, toResourcePath_js_1.toResourcePath)((0, common_1.isUrlAbsolute)(destUrl) ? destUrl : (0, common_1.combine)(uri.origin, destUrl)),
                                options: {
                                    KeepBoth: KeepBoth,
                                    ResetAuthorAndCreatedOnCopy: true,
                                    ShouldBypassSharedLocks: true,
                                    __metadata: {
                                        type: "SP.MoveCopyOptions",
                                    },
                                },
                                srcPath: (0, toResourcePath_js_1.toResourcePath)((0, common_1.combine)(uri.origin, urlInfo.Folder.ServerRelativeUrl)),
                            }))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Copies a folder to destination path
     *
     * @param destUrl Absolute or relative URL of the destination path
     */
    _Folder.prototype.copyTo = function (destUrl) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var urlInfo, uri;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getParentInfos()];
                    case 1:
                        urlInfo = _a.sent();
                        uri = new URL(urlInfo.ParentWeb.Url);
                        return [4 /*yield*/, (0, operations_js_1.spPost)((0, exports.Folder)(urlInfo.ParentWeb.Url, "/_api/SP.MoveCopyUtil.CopyFolder()").configureFrom(this), (0, odata_1.body)({
                                destUrl: (0, common_1.isUrlAbsolute)(destUrl) ? destUrl : (0, common_1.combine)(uri.origin, destUrl),
                                srcUrl: (0, common_1.combine)(uri.origin, urlInfo.Folder.ServerRelativeUrl),
                            }))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Copies a folder by path to destination path
     * Also works with different site collections.
     *
     * @param destUrl Absolute or relative URL of the destination path
     * @param keepBoth Keep both if folder with the same name in the same location already exists?
     */
    _Folder.prototype.copyByPath = function (destUrl, KeepBoth) {
        if (KeepBoth === void 0) { KeepBoth = false; }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var urlInfo, uri;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getParentInfos()];
                    case 1:
                        urlInfo = _a.sent();
                        uri = new URL(urlInfo.ParentWeb.Url);
                        return [4 /*yield*/, (0, operations_js_1.spPost)((0, exports.Folder)(uri.origin, "/_api/SP.MoveCopyUtil.CopyFolderByPath()").configureFrom(this), (0, odata_1.body)({
                                destPath: (0, toResourcePath_js_1.toResourcePath)((0, common_1.isUrlAbsolute)(destUrl) ? destUrl : (0, common_1.combine)(uri.origin, destUrl)),
                                options: {
                                    KeepBoth: KeepBoth,
                                    ResetAuthorAndCreatedOnCopy: true,
                                    ShouldBypassSharedLocks: true,
                                    __metadata: {
                                        type: "SP.MoveCopyOptions",
                                    },
                                },
                                srcPath: (0, toResourcePath_js_1.toResourcePath)((0, common_1.combine)(uri.origin, urlInfo.Folder.ServerRelativeUrl)),
                            }))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Deletes the folder object with options.
     *
     * @param parameters Specifies the options to use when deleting a folder.
     */
    _Folder.prototype.deleteWithParams = function (parameters) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                return [2 /*return*/, (0, operations_js_1.spPost)(this.clone(exports.Folder, "DeleteWithParameters"), (0, odata_1.body)({ parameters: parameters }))];
            });
        });
    };
    /**
     * Create the subfolder inside the current folder, as specified by the leafPath
     *
     * @param leafPath leafName of the new folder
     */
    _Folder.prototype.addSubFolderUsingPath = function (leafPath) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.spPost)(this.clone(exports.Folder, "AddSubFolderUsingPath"), (0, odata_1.body)({ leafPath: (0, toResourcePath_js_1.toResourcePath)(leafPath) }))];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, this.folders.getByName(leafPath)];
                }
            });
        });
    };
    /**
     * Gets the parent information for this folder's list and web
     */
    _Folder.prototype.getParentInfos = function () {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var urlInfo;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.select("ServerRelativeUrl", "ListItemAllFields/ParentList/Id", "ListItemAllFields/ParentList/RootFolder/UniqueId", "ListItemAllFields/ParentList/RootFolder/ServerRelativeUrl", "ListItemAllFields/ParentList/RootFolder/ServerRelativePath", "ListItemAllFields/ParentList/ParentWeb/Id", "ListItemAllFields/ParentList/ParentWeb/Url", "ListItemAllFields/ParentList/ParentWeb/ServerRelativeUrl", "ListItemAllFields/ParentList/ParentWeb/ServerRelativePath").expand("ListItemAllFields/ParentList", "ListItemAllFields/ParentList/RootFolder", "ListItemAllFields/ParentList/ParentWeb")()];
                    case 1:
                        urlInfo = _a.sent();
                        return [2 /*return*/, {
                                Folder: {
                                    ServerRelativeUrl: urlInfo.ServerRelativeUrl,
                                },
                                ParentList: {
                                    Id: urlInfo.ListItemAllFields.ParentList.Id,
                                    RootFolderServerRelativePath: urlInfo.ListItemAllFields.ParentList.RootFolder.ServerRelativePath,
                                    RootFolderServerRelativeUrl: urlInfo.ListItemAllFields.ParentList.RootFolder.ServerRelativeUrl,
                                    RootFolderUniqueId: urlInfo.ListItemAllFields.ParentList.RootFolder.UniqueId,
                                },
                                ParentWeb: {
                                    Id: urlInfo.ListItemAllFields.ParentList.ParentWeb.Id,
                                    ServerRelativePath: urlInfo.ListItemAllFields.ParentList.ParentWeb.ServerRelativePath,
                                    ServerRelativeUrl: urlInfo.ListItemAllFields.ParentList.ParentWeb.ServerRelativeUrl,
                                    Url: urlInfo.ListItemAllFields.ParentList.ParentWeb.Url,
                                },
                            }];
                }
            });
        });
    };
    /**
     * Gets the shareable item associated with this folder
     */
    _Folder.prototype.getShareable = function () {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var d, shareable;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.clone(sharepointqueryable_js_1.SharePointQueryableInstance, "listItemAllFields", false).select("odata.id")()];
                    case 1:
                        d = _a.sent();
                        shareable = (0, types_js_1.Item)((0, odata_js_1.odataUrlFrom)(d)).configureFrom(this);
                        // we need to handle batching
                        if (this.hasBatch) {
                            shareable = shareable.inBatch(this.batch);
                        }
                        return [2 /*return*/, shareable];
                }
            });
        });
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("f.recycle")
    ], _Folder.prototype, "recycle", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("f.getItem")
    ], _Folder.prototype, "getItem", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("f.moveTo")
    ], _Folder.prototype, "moveTo", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("f.moveByPath")
    ], _Folder.prototype, "moveByPath", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("f.copyTo")
    ], _Folder.prototype, "copyTo", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("f.copyByPath")
    ], _Folder.prototype, "copyByPath", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("f.del-params")
    ], _Folder.prototype, "deleteWithParams", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("f.getShareable")
    ], _Folder.prototype, "getShareable", null);
    return _Folder;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._Folder = _Folder;
exports.Folder = (0, sharepointqueryable_js_1.spInvokableFactory)(_Folder);
