// @pnp/sp-commonjs/user-custom-actions/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserCustomActionScope = exports.UserCustomActionRegistrationType = exports.UserCustomAction = exports._UserCustomAction = exports.UserCustomActions = exports._UserCustomActions = void 0;
var tslib_1 = require("tslib");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var common_1 = require("@pnp/common-commonjs");
var odata_1 = require("@pnp/odata-commonjs");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
var telemetry_js_1 = require("../telemetry.js");
var metadata_js_1 = require("../utils/metadata.js");
var _UserCustomActions = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_UserCustomActions, _super);
    function _UserCustomActions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Returns the user custom action with the specified id
     *
     * @param id The GUID id of the user custom action to retrieve
     */
    _UserCustomActions.prototype.getById = function (id) {
        return telemetry_js_1.tag.configure((0, exports.UserCustomAction)(this).concat("('" + id + "')"), "ucas.getById");
    };
    /**
     * Creates a user custom action
     *
     * @param properties The information object of property names and values which define the new user custom action
     */
    _UserCustomActions.prototype.add = function (properties) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.spPost)(this, (0, odata_1.body)((0, common_1.assign)((0, metadata_js_1.metadata)("SP.UserCustomAction"), properties)))];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                action: this.getById(data.Id),
                                data: data,
                            }];
                }
            });
        });
    };
    /**
     * Deletes all user custom actions in the collection
     */
    _UserCustomActions.prototype.clear = function () {
        return (0, operations_js_1.spPost)(this.clone(exports.UserCustomActions, "clear"));
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("ucas.add")
    ], _UserCustomActions.prototype, "add", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("ucas.clear")
    ], _UserCustomActions.prototype, "clear", null);
    _UserCustomActions = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("usercustomactions")
    ], _UserCustomActions);
    return _UserCustomActions;
}(sharepointqueryable_js_1._SharePointQueryableCollection));
exports._UserCustomActions = _UserCustomActions;
exports.UserCustomActions = (0, sharepointqueryable_js_1.spInvokableFactory)(_UserCustomActions);
var _UserCustomAction = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_UserCustomAction, _super);
    function _UserCustomAction() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.delete = (0, sharepointqueryable_js_1.deleteable)("uca");
        /**
        * Updates this user custom action with the supplied properties
        *
        * @param properties An information object of property names and values to update for this user custom action
        */
        _this.update = _this._update("SP.UserCustomAction", function (data) { return ({ data: data, action: _this }); });
        return _this;
    }
    return _UserCustomAction;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._UserCustomAction = _UserCustomAction;
exports.UserCustomAction = (0, sharepointqueryable_js_1.spInvokableFactory)(_UserCustomAction);
var UserCustomActionRegistrationType;
(function (UserCustomActionRegistrationType) {
    UserCustomActionRegistrationType[UserCustomActionRegistrationType["None"] = 0] = "None";
    UserCustomActionRegistrationType[UserCustomActionRegistrationType["List"] = 1] = "List";
    UserCustomActionRegistrationType[UserCustomActionRegistrationType["ContentType"] = 2] = "ContentType";
    UserCustomActionRegistrationType[UserCustomActionRegistrationType["ProgId"] = 3] = "ProgId";
    UserCustomActionRegistrationType[UserCustomActionRegistrationType["FileType"] = 4] = "FileType";
})(UserCustomActionRegistrationType = exports.UserCustomActionRegistrationType || (exports.UserCustomActionRegistrationType = {}));
var UserCustomActionScope;
(function (UserCustomActionScope) {
    UserCustomActionScope[UserCustomActionScope["Unknown"] = 0] = "Unknown";
    UserCustomActionScope[UserCustomActionScope["Site"] = 2] = "Site";
    UserCustomActionScope[UserCustomActionScope["Web"] = 3] = "Web";
    UserCustomActionScope[UserCustomActionScope["List"] = 4] = "List";
})(UserCustomActionScope = exports.UserCustomActionScope || (exports.UserCustomActionScope = {}));
