// @pnp/sp-commonjs/files/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Versions = exports.Version = exports.TemplateFileType = exports.MoveOperations = exports.CheckinType = exports.Files = exports.File = void 0;
require("./folder.js");
require("./item.js");
require("./web.js");
var types_js_1 = require("./types.js");
Object.defineProperty(exports, "File", { enumerable: true, get: function () { return types_js_1.File; } });
Object.defineProperty(exports, "Files", { enumerable: true, get: function () { return types_js_1.Files; } });
Object.defineProperty(exports, "CheckinType", { enumerable: true, get: function () { return types_js_1.CheckinType; } });
Object.defineProperty(exports, "MoveOperations", { enumerable: true, get: function () { return types_js_1.MoveOperations; } });
Object.defineProperty(exports, "TemplateFileType", { enumerable: true, get: function () { return types_js_1.TemplateFileType; } });
Object.defineProperty(exports, "Version", { enumerable: true, get: function () { return types_js_1.Version; } });
Object.defineProperty(exports, "Versions", { enumerable: true, get: function () { return types_js_1.Versions; } });
