// @pnp/odata-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disableExtensions = exports.enableExtensions = exports.clearGlobalExtensions = exports.extendFactory = exports.extendObj = exports.extendGlobal = void 0;
var tslib_1 = require("tslib");
(0, tslib_1.__exportStar)(require("./batch.js"), exports);
(0, tslib_1.__exportStar)(require("./caching.js"), exports);
(0, tslib_1.__exportStar)(require("./add-prop.js"), exports);
(0, tslib_1.__exportStar)(require("./invokable-binder.js"), exports);
(0, tslib_1.__exportStar)(require("./pipeline-binder.js"), exports);
(0, tslib_1.__exportStar)(require("./parsers.js"), exports);
(0, tslib_1.__exportStar)(require("./pipeline.js"), exports);
(0, tslib_1.__exportStar)(require("./queryable.js"), exports);
(0, tslib_1.__exportStar)(require("./request-builders.js"), exports);
var invokable_extensions_js_1 = require("./invokable-extensions.js");
Object.defineProperty(exports, "extendGlobal", { enumerable: true, get: function () { return invokable_extensions_js_1.extendGlobal; } });
Object.defineProperty(exports, "extendObj", { enumerable: true, get: function () { return invokable_extensions_js_1.extendObj; } });
Object.defineProperty(exports, "extendFactory", { enumerable: true, get: function () { return invokable_extensions_js_1.extendFactory; } });
Object.defineProperty(exports, "clearGlobalExtensions", { enumerable: true, get: function () { return invokable_extensions_js_1.clearGlobalExtensions; } });
Object.defineProperty(exports, "enableExtensions", { enumerable: true, get: function () { return invokable_extensions_js_1.enableExtensions; } });
Object.defineProperty(exports, "disableExtensions", { enumerable: true, get: function () { return invokable_extensions_js_1.disableExtensions; } });
