// @pnp/odata-commonjs/invokable-extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extensionOrDefault = exports.applyFactoryExtensions = exports.enableExtensions = exports.disableExtensions = exports.clearGlobalExtensions = exports.extendFactory = exports.extendObj = exports.extendGlobal = void 0;
var tslib_1 = require("tslib");
var common_1 = require("@pnp/common-commonjs");
var _enableExtensions = false;
var globalExtensions = [];
var factoryExtensions = new Map();
var ObjExtensionsSym = Symbol.for("43f7a601");
/**
 * Creates global extensions across all invokable objects
 *
 * @param e The global extensions to apply
 */
var extendGlobal = function (e) {
    _enableExtensions = true;
    extendCol(globalExtensions, e);
};
exports.extendGlobal = extendGlobal;
/**
 * Applies the supplied extensions to a single instance
 *
 * @param target Object to which extensions are applied
 * @param extensions Extensions to apply
 */
// eslint-disable-next-line @typescript-eslint/ban-types
var extendObj = function (target, extensions) {
    _enableExtensions = true;
    if (!Reflect.has(target, ObjExtensionsSym)) {
        Reflect.set(target, ObjExtensionsSym, []);
    }
    extendCol(Reflect.get(target, ObjExtensionsSym), extensions);
    return target;
};
exports.extendObj = extendObj;
/**
 * Allows applying extensions to all instances created from the supplied factory
 *
 * @param factory The Invokable Factory method to extend
 * @param extensions Extensions to apply
 */
var extendFactory = function (factory, extensions) {
    _enableExtensions = true;
    // factoryExtensions
    var proto = Reflect.getPrototypeOf(factory);
    if (!Reflect.has(proto, ObjExtensionsSym)) {
        Reflect.defineProperty(proto, ObjExtensionsSym, {
            value: (0, common_1.getGUID)(),
        });
    }
    var key = proto[ObjExtensionsSym];
    if (!factoryExtensions.has(key)) {
        factoryExtensions.set(key, []);
    }
    extendCol(factoryExtensions.get(key), extensions);
};
exports.extendFactory = extendFactory;
function extendCol(a, e) {
    if (Array.isArray(e)) {
        a.push.apply(a, (0, tslib_1.__spreadArray)([], (0, tslib_1.__read)(e), false));
    }
    else {
        a.push(e);
    }
}
/**
 * Clears all global extensions
 */
var clearGlobalExtensions = function () {
    globalExtensions.length = 0;
};
exports.clearGlobalExtensions = clearGlobalExtensions;
/**
 * Disables all extensions
 */
var disableExtensions = function () {
    _enableExtensions = false;
};
exports.disableExtensions = disableExtensions;
/**
 * Enables all extensions
 */
var enableExtensions = function () {
    _enableExtensions = true;
};
exports.enableExtensions = enableExtensions;
/**
 * Applies a set of extension previously applied to a factory using extendFactory to an object created from that factory
 *
 * @param factory
 * @param args
 */
// eslint-disable-next-line @typescript-eslint/ban-types
var applyFactoryExtensions = function (factory, args) {
    var o = factory(args);
    var proto = Reflect.getPrototypeOf(factory);
    if (Reflect.has(proto, ObjExtensionsSym)) {
        var extensions = factoryExtensions.get(Reflect.get(proto, ObjExtensionsSym));
        o = (0, exports.extendObj)(o, extensions);
    }
    return o;
};
exports.applyFactoryExtensions = applyFactoryExtensions;
function extensionOrDefault(op, or, target) {
    var rest = [];
    for (var _i = 3; _i < arguments.length; _i++) {
        rest[_i - 3] = arguments[_i];
    }
    if (_enableExtensions) {
        var extensions = [];
        // we need to first invoke extensions tied to only this object
        if (Reflect.has(target, ObjExtensionsSym)) {
            extensions.push.apply(extensions, (0, tslib_1.__spreadArray)([], (0, tslib_1.__read)(Reflect.get(target, ObjExtensionsSym)), false));
        }
        // second we need to process any global extensions
        extensions.push.apply(extensions, (0, tslib_1.__spreadArray)([], (0, tslib_1.__read)(globalExtensions), false));
        for (var i = 0; i < extensions.length; i++) {
            var extension = extensions[i];
            var result = undefined;
            if ((0, common_1.isFunc)(extension)) {
                // this extension is a function which we call
                result = extension.apply(void 0, (0, tslib_1.__spreadArray)([op, target], (0, tslib_1.__read)(rest), false));
            }
            else if (op === "get" && Reflect.has(extension, rest[0])) {
                // this extension is a named extension meaning we are overriding a specific method/property
                result = Reflect.get(extension, rest[0], target);
            }
            else if (Reflect.has(extension, op)) {
                // this extension is a ProxyHandler that has a handler defined for {op} so we pass control and see if we get a result
                result = Reflect.get(extension, op).apply(void 0, (0, tslib_1.__spreadArray)([target], (0, tslib_1.__read)(rest), false));
            }
            if (typeof result !== "undefined") {
                // if a extension returned a result, we return that
                // this means that this extension overrides any other extensions and no more are executed
                // first extension in the list to return "wins"
                return result;
            }
        }
    }
    return or.apply(void 0, (0, tslib_1.__spreadArray)([target], (0, tslib_1.__read)(rest), false));
}
exports.extensionOrDefault = extensionOrDefault;
