// @pnp/odata-commonjs/pipeline-binder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultPipelineBinder = exports.pipelineBinder = void 0;
var parsers_js_1 = require("./parsers.js");
var common_1 = require("@pnp/common-commonjs");
var queryable_js_1 = require("./queryable.js");
var pipeline_js_1 = require("./pipeline.js");
// first we bind the pipeline we will use for all requests within this closure
function pipelineBinder(pipes) {
    // then we bind the client factory we'll use (typically done in an implementing library such as sp)
    return function (clientFactory) {
        // then we create a binder we can apply for each type of method (GET, POST, etc.)
        return function (method) {
            // finally we get a function back to which we can pass an IQueryableData instance and execute the request it defines
            return function (o) {
                // send the IQueryableData down the pipeline
                return (0, pipeline_js_1.pipe)(Object.assign({}, {
                    batch: null,
                    batchDependency: null,
                    batchIndex: -1,
                    cachingOptions: null,
                    clientFactory: clientFactory,
                    cloneParentCacheOptions: null,
                    cloneParentWasCaching: false,
                    hasResult: false,
                    isBatched: (0, common_1.objectDefinedNotNull)(o.batch),
                    method: method,
                    options: null,
                    parentUrl: "",
                    parser: new parsers_js_1.ODataParser(),
                    pipes: pipes.slice(0),
                    query: new Map(),
                    requestId: (0, common_1.getGUID)(),
                    url: "",
                    useCaching: /^get$/i.test(o.method) && o.useCaching,
                }, (0, queryable_js_1.cloneQueryableData)(o)));
            };
        };
    };
}
exports.pipelineBinder = pipelineBinder;
exports.defaultPipelineBinder = pipelineBinder((0, pipeline_js_1.getDefaultPipeline)());
