// @pnp/sp-commonjs/sharing/file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var types_js_1 = require("../files/types.js");
var types_js_2 = require("./types.js");
var splibconfig_js_1 = require("../splibconfig.js");
types_js_1._File.prototype.shareWith = function (loginNames, role, requireSignin, emailData) {
    if (role === void 0) { role = types_js_2.SharingRole.View; }
    if (requireSignin === void 0) { requireSignin = false; }
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.shareWith(loginNames, role, requireSignin, emailData)];
            }
        });
    });
};
types_js_1._File.prototype.getShareLink = function (kind, expiration) {
    if (expiration === void 0) { expiration = null; }
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.getShareLink(kind, expiration)];
            }
        });
    });
};
types_js_1._File.prototype.checkSharingPermissions = function (recipients) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.checkSharingPermissions(recipients)];
            }
        });
    });
};
// eslint-disable-next-line max-len
types_js_1._File.prototype.getSharingInformation = function (request, expands) {
    if (request === void 0) { request = null; }
    if (expands === void 0) { expands = []; }
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.getSharingInformation(request, expands)];
            }
        });
    });
};
types_js_1._File.prototype.getObjectSharingSettings = function (useSimplifiedRoles) {
    if (useSimplifiedRoles === void 0) { useSimplifiedRoles = true; }
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.getObjectSharingSettings(useSimplifiedRoles)];
            }
        });
    });
};
types_js_1._File.prototype.unshare = function () {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.unshare()];
            }
        });
    });
};
types_js_1._File.prototype.deleteSharingLinkByKind = function (linkKind) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.deleteSharingLinkByKind(linkKind)];
            }
        });
    });
};
types_js_1._File.prototype.unshareLink = function unshareLink(linkKind, shareId) {
    if (shareId === void 0) { shareId = splibconfig_js_1.emptyGuid; }
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var item;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this.getItem()];
                case 1:
                    item = _a.sent();
                    return [2 /*return*/, item.unshareLink(linkKind, shareId)];
            }
        });
    });
};
