// @pnp/sp-commonjs/sharing/item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var types_js_1 = require("../items/types.js");
var types_js_2 = require("./types.js");
var funcs_js_1 = require("./funcs.js");
types_js_1._Item.prototype.shareWith = function (loginNames, role, requireSignin, emailData) {
    if (role === void 0) { role = types_js_2.SharingRole.View; }
    if (requireSignin === void 0) { requireSignin = false; }
    return (0, funcs_js_1.shareWith)(this, loginNames, role, requireSignin, false, emailData);
};
types_js_1._Item.prototype.getShareLink = funcs_js_1.getShareLink;
types_js_1._Item.prototype.checkSharingPermissions = funcs_js_1.checkPermissions;
types_js_1._Item.prototype.getSharingInformation = funcs_js_1.getSharingInformation;
types_js_1._Item.prototype.getObjectSharingSettings = funcs_js_1.getObjectSharingSettings;
types_js_1._Item.prototype.unshare = funcs_js_1.unshareObject;
types_js_1._Item.prototype.deleteSharingLinkByKind = funcs_js_1.deleteLinkByKind;
types_js_1._Item.prototype.unshareLink = funcs_js_1.unshareLink;
