// @pnp/sp-commonjs/sites/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Site = exports._Site = void 0;
var tslib_1 = require("tslib");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var decorators_js_1 = require("../decorators.js");
var types_js_1 = require("../webs/types.js");
var common_1 = require("@pnp/common-commonjs");
var odata_1 = require("@pnp/odata-commonjs");
var odata_js_1 = require("../odata.js");
var operations_js_1 = require("../operations.js");
var batch_js_1 = require("../batch.js");
var escapeQueryStrValue_js_1 = require("../utils/escapeQueryStrValue.js");
var telemetry_js_1 = require("../telemetry.js");
var metadata_js_1 = require("../utils/metadata.js");
var extractweburl_js_1 = require("../utils/extractweburl.js");
var splibconfig_js_1 = require("../splibconfig.js");
var _Site = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Site, _super);
    function _Site() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(_Site.prototype, "rootWeb", {
        /**
         * Gets the root web of the site collection
         *
         */
        get: function () {
            return telemetry_js_1.tag.configure((0, types_js_1.Web)(this, "rootweb").configureFrom(this), "si.rootWeb");
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Returns the collection of changes from the change log that have occurred within the list, based on the specified query
     *
     * @param query The change query
     */
    _Site.prototype.getChanges = function (query) {
        var postBody = (0, odata_1.body)({ "query": (0, common_1.assign)((0, metadata_js_1.metadata)("SP.ChangeQuery"), query) });
        return (0, operations_js_1.spPost)(this.clone(types_js_1.Web, "getchanges"), postBody);
    };
    /**
     * Opens a web by id (using POST)
     *
     * @param webId The GUID id of the web to open
     */
    _Site.prototype.openWebById = function (webId) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.spPost)(this.clone(exports.Site, "openWebById('" + webId + "')"))];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                web: (0, types_js_1.Web)((0, extractweburl_js_1.extractWebUrl)((0, odata_js_1.odataUrlFrom)(data))).configureFrom(this),
                            }];
                }
            });
        });
    };
    /**
     * Gets a Web instance representing the root web of the site collection
     * correctly setup for chaining within the library
     */
    _Site.prototype.getRootWeb = function () {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var web;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.rootWeb.select("Url")()];
                    case 1:
                        web = _a.sent();
                        return [2 /*return*/, telemetry_js_1.tag.configure((0, types_js_1.Web)(web.Url).configureFrom(this), "si.getRootWeb")];
                }
            });
        });
    };
    /**
     * Gets the context information for this site collection
     */
    _Site.prototype.getContextInfo = function () {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var site, q, data, info;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        site = (0, exports.Site)(this.parentUrl, "_api/contextinfo").configureFrom(this);
                        q = telemetry_js_1.tag.configure(site, "si.getContextInfo");
                        return [4 /*yield*/, (0, operations_js_1.spPost)(q)];
                    case 1:
                        data = _a.sent();
                        if ((0, common_1.hOP)(data, "GetContextWebInformation")) {
                            info = data.GetContextWebInformation;
                            info.SupportedSchemaVersions = info.SupportedSchemaVersions.results;
                            return [2 /*return*/, info];
                        }
                        else {
                            return [2 /*return*/, data];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    _Site.prototype.createBatch = function () {
        return new batch_js_1.SPBatch(this.parentUrl, this.getRuntime());
    };
    /**
     * Deletes the current site
     *
     */
    _Site.prototype.delete = function () {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var site, q;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.clone(exports.Site, "").select("Id")()];
                    case 1:
                        site = _a.sent();
                        q = telemetry_js_1.tag.configure((0, exports.Site)(this.parentUrl, "_api/SPSiteManager/Delete"), "si.delete");
                        return [4 /*yield*/, (0, operations_js_1.spPost)(q, (0, odata_1.body)({ siteId: site.Id }))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Gets the document libraries on a site. Static method. (SharePoint Online only)
     *
     * @param absoluteWebUrl The absolute url of the web whose document libraries should be returned
     */
    _Site.prototype.getDocumentLibraries = function (absoluteWebUrl) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var q, data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        q = telemetry_js_1.tag.configure((0, sharepointqueryable_js_1.SharePointQueryable)("", "_api/sp.web.getdocumentlibraries(@v)"), "si.getDocumentLibraries");
                        q.query.set("@v", "'" + (0, escapeQueryStrValue_js_1.escapeQueryStrValue)(absoluteWebUrl) + "'");
                        return [4 /*yield*/, q()];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, (0, common_1.hOP)(data, "GetDocumentLibraries") ? data.GetDocumentLibraries : data];
                }
            });
        });
    };
    /**
     * Gets the site url from a page url
     *
     * @param absolutePageUrl The absolute url of the page
     */
    _Site.prototype.getWebUrlFromPageUrl = function (absolutePageUrl) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var q, data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        q = telemetry_js_1.tag.configure((0, sharepointqueryable_js_1.SharePointQueryable)("", "_api/sp.web.getweburlfrompageurl(@v)"), "si.getWebUrlFromPageUrl");
                        q.query.set("@v", "'" + (0, escapeQueryStrValue_js_1.escapeQueryStrValue)(absolutePageUrl) + "'");
                        return [4 /*yield*/, q()];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, (0, common_1.hOP)(data, "GetWebUrlFromPageUrl") ? data.GetWebUrlFromPageUrl : data];
                }
            });
        });
    };
    /**
     * Creates a Modern communication site.
     *
     * @param title The title of the site to create
     * @param lcid The language to use for the site. If not specified will default to 1033 (English).
     * @param shareByEmailEnabled If set to true, it will enable sharing files via Email. By default it is set to false
     * @param url The fully qualified URL (e.g. https://yourtenant.sharepoint.com/sites/mysitecollection) of the site.
     * @param description The description of the communication site.
     * @param classification The Site classification to use. For instance 'Contoso Classified'. See https://www.youtube.com/watch?v=E-8Z2ggHcS0 for more information
     * @param siteDesignId The Guid of the site design to be used.
     *                     You can use the below default OOTB GUIDs:
     *                     Topic: 00000000-0000-0000-0000-000000000000
     *                     Showcase: 6142d2a0-63a5-4ba0-aede-d9fefca2c767
     *                     Blank: f6cc5403-0d63-442e-96c0-285923709ffc
     * @param hubSiteId The id of the hub site to which the new site should be associated
     * @param owner Optional owner value, required if executing the method in app only mode
     */
    _Site.prototype.createCommunicationSite = function (title, lcid, shareByEmailEnabled, url, description, classification, siteDesignId, hubSiteId, owner) {
        if (lcid === void 0) { lcid = 1033; }
        if (shareByEmailEnabled === void 0) { shareByEmailEnabled = false; }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                return [2 /*return*/, this.createCommunicationSiteFromProps({
                        Classification: classification,
                        Description: description,
                        HubSiteId: hubSiteId,
                        Lcid: lcid,
                        Owner: owner,
                        ShareByEmailEnabled: shareByEmailEnabled,
                        SiteDesignId: siteDesignId,
                        Title: title,
                        Url: url,
                    })];
            });
        });
    };
    _Site.prototype.createCommunicationSiteFromProps = function (props) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var p, postBody;
            return (0, tslib_1.__generator)(this, function (_a) {
                p = Object.assign({}, {
                    Classification: "",
                    Description: "",
                    HubSiteId: splibconfig_js_1.emptyGuid,
                    Lcid: 1033,
                    ShareByEmailEnabled: false,
                    SiteDesignId: splibconfig_js_1.emptyGuid,
                    WebTemplate: "SITEPAGEPUBLISHING#0",
                    WebTemplateExtensionId: splibconfig_js_1.emptyGuid,
                }, props);
                postBody = (0, odata_1.body)({
                    "request": (0, common_1.assign)((0, metadata_js_1.metadata)("Microsoft.SharePoint.Portal.SPSiteCreationRequest"), p),
                });
                return [2 /*return*/, (0, operations_js_1.spPost)((0, exports.Site)((0, extractweburl_js_1.extractWebUrl)(this.toUrl()), "/_api/SPSiteManager/Create"), postBody)];
            });
        });
    };
    /**
     *
     * @param url Site Url that you want to check if exists
     */
    _Site.prototype.exists = function (url) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var postBody, value;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        postBody = (0, odata_1.body)({ url: url });
                        return [4 /*yield*/, (0, operations_js_1.spPost)((0, exports.Site)((0, extractweburl_js_1.extractWebUrl)(this.toUrl()), "/_api/SP.Site.Exists"), postBody)];
                    case 1:
                        value = _a.sent();
                        return [2 /*return*/, value];
                }
            });
        });
    };
    /**
     * Creates a Modern team site backed by Office 365 group. For use in SP Online only. This will not work with App-only tokens
     *
     * @param displayName The title or display name of the Modern team site to be created
     * @param alias Alias of the underlying Office 365 Group
     * @param isPublic Defines whether the Office 365 Group will be public (default), or private.
     * @param lcid The language to use for the site. If not specified will default to English (1033).
     * @param description The description of the site to be created.
     * @param classification The Site classification to use. For instance 'Contoso Classified'. See https://www.youtube.com/watch?v=E-8Z2ggHcS0 for more information
     * @param owners The Owners of the site to be created
     */
    _Site.prototype.createModernTeamSite = function (displayName, alias, isPublic, lcid, description, classification, owners, hubSiteId, siteDesignId) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                return [2 /*return*/, this.createModernTeamSiteFromProps({
                        alias: alias,
                        classification: classification,
                        description: description,
                        displayName: displayName,
                        hubSiteId: hubSiteId,
                        isPublic: isPublic,
                        lcid: lcid,
                        owners: owners,
                        siteDesignId: siteDesignId,
                    })];
            });
        });
    };
    _Site.prototype.createModernTeamSiteFromProps = function (props) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var p, postBody;
            return (0, tslib_1.__generator)(this, function (_a) {
                p = Object.assign({}, {
                    classification: "",
                    description: "",
                    hubSiteId: splibconfig_js_1.emptyGuid,
                    isPublic: true,
                    lcid: 1033,
                    owners: [],
                }, props);
                postBody = {
                    alias: p.alias,
                    displayName: p.displayName,
                    isPublic: p.isPublic,
                    optionalParams: {
                        Classification: p.classification,
                        CreationOptions: {
                            "results": ["SPSiteLanguage:" + p.lcid, "HubSiteId:" + p.hubSiteId],
                        },
                        Description: p.description,
                        Owners: {
                            "results": p.owners,
                        },
                    },
                };
                if (p.siteDesignId) {
                    postBody.optionalParams.CreationOptions.results.push("implicit_formula_292aa8a00786498a87a5ca52d9f4214a_" + p.siteDesignId);
                }
                return [2 /*return*/, (0, operations_js_1.spPost)((0, exports.Site)((0, extractweburl_js_1.extractWebUrl)(this.toUrl()), "/_api/GroupSiteManager/CreateGroupEx"), (0, odata_1.body)(postBody))];
            });
        });
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("si.getChanges")
    ], _Site.prototype, "getChanges", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("si.openWebById")
    ], _Site.prototype, "openWebById", null);
    _Site = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("_api/site")
    ], _Site);
    return _Site;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._Site = _Site;
exports.Site = (0, sharepointqueryable_js_1.spInvokableFactory)(_Site);
