// @pnp/sp-commonjs/search/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Suggest = exports.Search = exports.SearchResults = exports.SearchQueryBuilder = void 0;
var tslib_1 = require("tslib");
var rest_js_1 = require("../rest.js");
var query_js_1 = require("./query.js");
var query_js_2 = require("./query.js");
var suggest_js_1 = require("./suggest.js");
(0, tslib_1.__exportStar)(require("./types.js"), exports);
var query_js_3 = require("./query.js");
Object.defineProperty(exports, "SearchQueryBuilder", { enumerable: true, get: function () { return query_js_3.SearchQueryBuilder; } });
Object.defineProperty(exports, "SearchResults", { enumerable: true, get: function () { return query_js_3.SearchResults; } });
Object.defineProperty(exports, "Search", { enumerable: true, get: function () { return query_js_3.Search; } });
var suggest_js_2 = require("./suggest.js");
Object.defineProperty(exports, "Suggest", { enumerable: true, get: function () { return suggest_js_2.Suggest; } });
rest_js_1.SPRest.prototype.search = function (query) {
    return this.childConfigHook(function (_a) {
        var options = _a.options, baseUrl = _a.baseUrl, runtime = _a.runtime;
        return (0, query_js_2.Search)(baseUrl, options, runtime)(query);
    });
};
rest_js_1.SPRest.prototype.searchWithCaching = function (query, cacheOptions) {
    return this.childConfigHook(function (_a) {
        var options = _a.options, baseUrl = _a.baseUrl, runtime = _a.runtime;
        return (new query_js_1._Search(baseUrl)).configure(options).setRuntime(runtime).usingCaching(cacheOptions).execute(query);
    });
};
rest_js_1.SPRest.prototype.searchSuggest = function (query) {
    return this.childConfigHook(function (_a) {
        var options = _a.options, baseUrl = _a.baseUrl, runtime = _a.runtime;
        return (0, suggest_js_1.Suggest)(baseUrl, options, runtime)(typeof query === "string" ? { querytext: query } : query);
    });
};
