// @pnp/sp-commonjs/search/suggest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Suggest = exports._Suggest = void 0;
var tslib_1 = require("tslib");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var common_1 = require("@pnp/common-commonjs");
var decorators_js_1 = require("../decorators.js");
var telemetry_js_1 = require("../telemetry.js");
var _Suggest = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Suggest, _super);
    function _Suggest() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    _Suggest.prototype.execute = function (query) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var response, mapper;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.mapQueryToQueryString(query);
                        return [4 /*yield*/, this.get()];
                    case 1:
                        response = _a.sent();
                        mapper = (0, common_1.hOP)(response, "suggest") ? function (s_1) { return response.suggest[s_1].results; } : function (s_2) { return response[s_2]; };
                        return [2 /*return*/, {
                                PeopleNames: mapper("PeopleNames"),
                                PersonalResults: mapper("PersonalResults"),
                                Queries: mapper("Queries"),
                            }];
                }
            });
        });
    };
    _Suggest.prototype.mapQueryToQueryString = function (query) {
        var _this = this;
        var setProp = function (q) { return function (checkProp) { return function (sp) {
            if ((0, common_1.hOP)(q, checkProp)) {
                _this.query.set(sp, q[checkProp].toString());
            }
        }; }; };
        this.query.set("querytext", "'" + query.querytext + "'");
        var querySetter = setProp(query);
        querySetter("count")("inumberofquerysuggestions");
        querySetter("personalCount")("inumberofresultsuggestions");
        querySetter("preQuery")("fprequerysuggestions");
        querySetter("hitHighlighting")("fhithighlighting");
        querySetter("capitalize")("fcapitalizefirstletters");
        querySetter("culture")("culture");
        querySetter("stemming")("enablestemming");
        querySetter("includePeople")("showpeoplenamesuggestions");
        querySetter("queryRules")("enablequeryrules");
        querySetter("prefixMatch")("fprefixmatchallterms");
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("su.execute")
    ], _Suggest.prototype, "execute", null);
    _Suggest = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("_api/search/suggest")
    ], _Suggest);
    return _Suggest;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._Suggest = _Suggest;
var Suggest = function (baseUrl, options, runtime) {
    if (options === void 0) { options = {}; }
    if (runtime === void 0) { runtime = common_1.DefaultRuntime; }
    return function (query) {
        return (new _Suggest(baseUrl)).configure(options).setRuntime(runtime).execute(query);
    };
};
exports.Suggest = Suggest;
