// @pnp/sp-commonjs/views/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewScope = exports.ViewFields = exports._ViewFields = exports.View = exports._View = exports.Views = exports._Views = void 0;
var tslib_1 = require("tslib");
var odata_1 = require("@pnp/odata-commonjs");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var metadata_js_1 = require("../utils/metadata.js");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
var telemetry_js_1 = require("../telemetry.js");
var _Views = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Views, _super);
    function _Views() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Adds a new view to the collection
     *
     * @param title The new views's title
     * @param personalView True if this is a personal view, otherwise false, default = false
     * @param additionalSettings Will be passed as part of the view creation body
     */
    _Views.prototype.add = function (title, personalView, additionalSettings) {
        if (personalView === void 0) { personalView = false; }
        if (additionalSettings === void 0) { additionalSettings = {}; }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var postBody, data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        postBody = (0, odata_1.body)(Object.assign((0, metadata_js_1.metadata)("SP.View"), {
                            "PersonalView": personalView,
                            "Title": title,
                        }, additionalSettings));
                        return [4 /*yield*/, (0, operations_js_1.spPost)(this.clone(exports.Views, null), postBody)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                view: this.getById(data.Id),
                            }];
                }
            });
        });
    };
    /**
     * Gets a view by guid id
     *
     * @param id The GUID id of the view
     */
    _Views.prototype.getById = function (id) {
        return (0, exports.View)(this).concat("('" + id + "')");
    };
    /**
     * Gets a view by title (case-sensitive)
     *
     * @param title The case-sensitive title of the view
     */
    _Views.prototype.getByTitle = function (title) {
        return (0, exports.View)(this, "getByTitle('" + title + "')");
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("vs.add")
    ], _Views.prototype, "add", null);
    _Views = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("views")
    ], _Views);
    return _Views;
}(sharepointqueryable_js_1._SharePointQueryableCollection));
exports._Views = _Views;
exports.Views = (0, sharepointqueryable_js_1.spInvokableFactory)(_Views);
var _View = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_View, _super);
    function _View() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.delete = (0, sharepointqueryable_js_1.deleteable)("vw");
        /**
         * Updates this view intance with the supplied properties
         *
         * @param properties A plain object hash of values to update for the view
         */
        _this.update = _this._update("SP.View", function (data) { return ({ data: data, view: _this }); });
        return _this;
    }
    Object.defineProperty(_View.prototype, "fields", {
        get: function () {
            return (0, exports.ViewFields)(this);
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Returns the list view as HTML.
     *
     */
    _View.prototype.renderAsHtml = function () {
        return this.clone(exports.View, "renderashtml")();
    };
    /**
     * Sets the view schema
     *
     * @param viewXml The view XML to set
     */
    _View.prototype.setViewXml = function (viewXml) {
        return (0, operations_js_1.spPost)(this.clone(exports.View, "SetViewXml"), (0, odata_1.body)({ viewXml: viewXml }));
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("v.renderAsHtml")
    ], _View.prototype, "renderAsHtml", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("v.setViewXml")
    ], _View.prototype, "setViewXml", null);
    return _View;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._View = _View;
exports.View = (0, sharepointqueryable_js_1.spInvokableFactory)(_View);
var _ViewFields = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_ViewFields, _super);
    function _ViewFields() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Gets a value that specifies the XML schema that represents the collection.
     */
    _ViewFields.prototype.getSchemaXml = function () {
        return this.clone(exports.ViewFields, "schemaxml")();
    };
    /**
     * Adds the field with the specified field internal name or display name to the collection.
     *
     * @param fieldTitleOrInternalName The case-sensitive internal name or display name of the field to add.
     */
    _ViewFields.prototype.add = function (fieldTitleOrInternalName) {
        return (0, operations_js_1.spPost)(this.clone(exports.ViewFields, "addviewfield('" + fieldTitleOrInternalName + "')"));
    };
    /**
     * Moves the field with the specified field internal name to the specified position in the collection.
     *
     * @param field The case-sensitive internal name of the field to move.
     * @param index The zero-based index of the new position for the field.
     */
    _ViewFields.prototype.move = function (field, index) {
        return (0, operations_js_1.spPost)(this.clone(exports.ViewFields, "moveviewfieldto"), (0, odata_1.body)({ field: field, index: index }));
    };
    /**
     * Removes all the fields from the collection.
     */
    _ViewFields.prototype.removeAll = function () {
        return (0, operations_js_1.spPost)(this.clone(exports.ViewFields, "removeallviewfields"));
    };
    /**
     * Removes the field with the specified field internal name from the collection.
     *
     * @param fieldInternalName The case-sensitive internal name of the field to remove from the view.
     */
    _ViewFields.prototype.remove = function (fieldInternalName) {
        return (0, operations_js_1.spPost)(this.clone(exports.ViewFields, "removeviewfield('" + fieldInternalName + "')"));
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("vfs.getSchemaXml")
    ], _ViewFields.prototype, "getSchemaXml", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("vfs.add")
    ], _ViewFields.prototype, "add", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("vfs.move")
    ], _ViewFields.prototype, "move", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("vfs.removeAll")
    ], _ViewFields.prototype, "removeAll", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("vfs.remove")
    ], _ViewFields.prototype, "remove", null);
    _ViewFields = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("viewfields")
    ], _ViewFields);
    return _ViewFields;
}(sharepointqueryable_js_1._SharePointQueryableCollection));
exports._ViewFields = _ViewFields;
exports.ViewFields = (0, sharepointqueryable_js_1.spInvokableFactory)(_ViewFields);
var ViewScope;
(function (ViewScope) {
    ViewScope[ViewScope["DefaultValue"] = 0] = "DefaultValue";
    ViewScope[ViewScope["Recursive"] = 1] = "Recursive";
    ViewScope[ViewScope["RecursiveAll"] = 2] = "RecursiveAll";
    ViewScope[ViewScope["FilesOnly"] = 3] = "FilesOnly";
})(ViewScope = exports.ViewScope || (exports.ViewScope = {}));
