// @pnp/sp-commonjs/related-items/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RelatedItemManager = exports._RelatedItemManager = void 0;
var tslib_1 = require("tslib");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var extractweburl_js_1 = require("../utils/extractweburl.js");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
var odata_1 = require("@pnp/odata-commonjs");
var telemetry_js_1 = require("../telemetry.js");
var _RelatedItemManager = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_RelatedItemManager, _super);
    function _RelatedItemManager() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    _RelatedItemManager.prototype.getRelatedItems = function (sourceListName, sourceItemId) {
        var query = this.clone(exports.RelatedItemManager, null);
        query.concat(".GetRelatedItems");
        return (0, operations_js_1.spPost)(query, (0, odata_1.body)({
            SourceItemID: sourceItemId,
            SourceListName: sourceListName,
        }));
    };
    _RelatedItemManager.prototype.getPageOneRelatedItems = function (sourceListName, sourceItemId) {
        var query = this.clone(exports.RelatedItemManager, null);
        query.concat(".GetPageOneRelatedItems");
        return (0, operations_js_1.spPost)(query, (0, odata_1.body)({
            SourceItemID: sourceItemId,
            SourceListName: sourceListName,
        }));
    };
    _RelatedItemManager.prototype.addSingleLink = function (sourceListName, sourceItemId, sourceWebUrl, targetListName, targetItemID, targetWebUrl, tryAddReverseLink) {
        if (tryAddReverseLink === void 0) { tryAddReverseLink = false; }
        var query = this.clone(exports.RelatedItemManager, null);
        query.concat(".AddSingleLink");
        return (0, operations_js_1.spPost)(query, (0, odata_1.body)({
            SourceItemID: sourceItemId,
            SourceListName: sourceListName,
            SourceWebUrl: sourceWebUrl,
            TargetItemID: targetItemID,
            TargetListName: targetListName,
            TargetWebUrl: targetWebUrl,
            TryAddReverseLink: tryAddReverseLink,
        }));
    };
    _RelatedItemManager.prototype.addSingleLinkToUrl = function (sourceListName, sourceItemId, targetItemUrl, tryAddReverseLink) {
        if (tryAddReverseLink === void 0) { tryAddReverseLink = false; }
        var query = this.clone(exports.RelatedItemManager, null);
        query.concat(".AddSingleLinkToUrl");
        return (0, operations_js_1.spPost)(query, (0, odata_1.body)({
            SourceItemID: sourceItemId,
            SourceListName: sourceListName,
            TargetItemUrl: targetItemUrl,
            TryAddReverseLink: tryAddReverseLink,
        }));
    };
    _RelatedItemManager.prototype.addSingleLinkFromUrl = function (sourceItemUrl, targetListName, targetItemId, tryAddReverseLink) {
        if (tryAddReverseLink === void 0) { tryAddReverseLink = false; }
        var query = this.clone(exports.RelatedItemManager, null);
        query.concat(".AddSingleLinkFromUrl");
        return (0, operations_js_1.spPost)(query, (0, odata_1.body)({
            SourceItemUrl: sourceItemUrl,
            TargetItemID: targetItemId,
            TargetListName: targetListName,
            TryAddReverseLink: tryAddReverseLink,
        }));
    };
    _RelatedItemManager.prototype.deleteSingleLink = function (sourceListName, sourceItemId, sourceWebUrl, targetListName, targetItemId, targetWebUrl, tryDeleteReverseLink) {
        if (tryDeleteReverseLink === void 0) { tryDeleteReverseLink = false; }
        var query = this.clone(exports.RelatedItemManager, null);
        query.concat(".DeleteSingleLink");
        return (0, operations_js_1.spPost)(query, (0, odata_1.body)({
            SourceItemID: sourceItemId,
            SourceListName: sourceListName,
            SourceWebUrl: sourceWebUrl,
            TargetItemID: targetItemId,
            TargetListName: targetListName,
            TargetWebUrl: targetWebUrl,
            TryDeleteReverseLink: tryDeleteReverseLink,
        }));
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("rim.getRelatedItems")
    ], _RelatedItemManager.prototype, "getRelatedItems", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("rim.getPageOneRelatedItems")
    ], _RelatedItemManager.prototype, "getPageOneRelatedItems", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("rim.addSingleLink")
    ], _RelatedItemManager.prototype, "addSingleLink", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("rim.ToUrl")
    ], _RelatedItemManager.prototype, "addSingleLinkToUrl", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("rim.FromUrl")
    ], _RelatedItemManager.prototype, "addSingleLinkFromUrl", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("rim.deleteSingleLink")
    ], _RelatedItemManager.prototype, "deleteSingleLink", null);
    _RelatedItemManager = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("_api/SP.RelatedItemManager")
    ], _RelatedItemManager);
    return _RelatedItemManager;
}(sharepointqueryable_js_1._SharePointQueryable));
exports._RelatedItemManager = _RelatedItemManager;
var RelatedItemManager = function (url) { return new _RelatedItemManager((0, extractweburl_js_1.extractWebUrl)(typeof url === "string" ? url : url.toUrl())); };
exports.RelatedItemManager = RelatedItemManager;
