// @pnp/sp-commonjs/features/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Feature = exports._Feature = exports.Features = exports._Features = void 0;
var tslib_1 = require("tslib");
var odata_1 = require("@pnp/odata-commonjs");
var sharepointqueryable_js_1 = require("../sharepointqueryable.js");
var decorators_js_1 = require("../decorators.js");
var operations_js_1 = require("../operations.js");
var telemetry_js_1 = require("../telemetry.js");
var _Features = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Features, _super);
    function _Features() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Adds (activates) the specified feature
     *
     * @param id The Id of the feature (GUID)
     * @param force If true the feature activation will be forced
     */
    _Features.prototype.add = function (id, force) {
        if (force === void 0) { force = false; }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, operations_js_1.spPost)(this.clone(exports.Features, "add"), (0, odata_1.body)({
                            featdefScope: 0,
                            featureId: id,
                            force: force,
                        }))];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, {
                                data: data,
                                feature: this.getById(id),
                            }];
                }
            });
        });
    };
    /**
     * Gets a feature from the collection with the specified guid
     *
     * @param id The Id of the feature (GUID)
     */
    _Features.prototype.getById = function (id) {
        var feature = (0, exports.Feature)(this);
        feature.concat("('" + id + "')");
        return telemetry_js_1.tag.configure(feature, "fes.getById");
    };
    /**
     * Removes (deactivates) a feature from the collection
     *
     * @param id The Id of the feature (GUID)
     * @param force If true the feature deactivation will be forced
     */
    _Features.prototype.remove = function (id, force) {
        if (force === void 0) { force = false; }
        return (0, operations_js_1.spPost)(this.clone(exports.Features, "remove"), (0, odata_1.body)({
            featureId: id,
            force: force,
        }));
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("fes.add")
    ], _Features.prototype, "add", null);
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("fes.remove")
    ], _Features.prototype, "remove", null);
    _Features = (0, tslib_1.__decorate)([
        (0, decorators_js_1.defaultPath)("features")
    ], _Features);
    return _Features;
}(sharepointqueryable_js_1._SharePointQueryableCollection));
exports._Features = _Features;
exports.Features = (0, sharepointqueryable_js_1.spInvokableFactory)(_Features);
var _Feature = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(_Feature, _super);
    function _Feature() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Removes (deactivates) the feature
     *
     * @param force If true the feature deactivation will be forced
     */
    _Feature.prototype.deactivate = function (force) {
        if (force === void 0) { force = false; }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var removeDependency, feature, promise;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        removeDependency = this.addBatchDependency();
                        return [4 /*yield*/, (0, exports.Feature)(this).select("DefinitionId")()];
                    case 1:
                        feature = _a.sent();
                        promise = this.getParent(exports.Features, this.parentUrl, "", this.batch).remove(feature.DefinitionId, force);
                        removeDependency();
                        return [2 /*return*/, promise];
                }
            });
        });
    };
    (0, tslib_1.__decorate)([
        (0, telemetry_js_1.tag)("fe.deactivate")
    ], _Feature.prototype, "deactivate", null);
    return _Feature;
}(sharepointqueryable_js_1._SharePointQueryableInstance));
exports._Feature = _Feature;
exports.Feature = (0, sharepointqueryable_js_1.spInvokableFactory)(_Feature);
